art/estimators/speech_recognition/tensorflow_lingvo.py

```diff
@@ -0,0 +1,516 @@
+        """
+        if x[0].ndim != 1:
+            raise ValueError(
+                "The LingvoAsr estimator can only be used temporal data of type mono. Please remove any channel"
+                "dimension."
+            )
+
+        nb_samples = x.shape[0]
+        assert nb_samples % batch_size == 0, "Number of samples must be divisible by batch_size"
+
+        y = list()
+        nb_batches = int(np.ceil(nb_samples / float(batch_size)))
+        for m in range(nb_batches):
+            # batch indices
+            begin, end = m * batch_size, min((m + 1) * batch_size, nb_samples)
+
+            x_batch_padded, _, mask_frequency = self._pad_audio_input(x[begin:end])
+
+            feed_dict = {
+                self._x_padded: x_batch_padded,
+                self._y_target: np.array(["DUMMY"] * batch_size),
+                self._mask_frequency: mask_frequency,
+            }
+            # run prediction
+            y_batch = self._sess.run(self._predict_batch_op, feed_dict)
+
+            # extract and append transcription result
+            y += y_batch["topk_decoded"][:, 0].tolist()
+
+        y_decoded = [item.decode("utf-8") for item in y]
+        return np.array(y_decoded, dtype=str)
+
+    def _loss_gradient(self, x: "Tensor", y: "Tensor", mask: "Tensor") -> "Tensor":
+        """Define loss gradients computation operation for a batch of padded inputs."""
+        import tensorflow.compat.v1 as tf1
+
+        # create decoder inputs
+        decoder_inputs = self._create_decoder_input(x, y, mask)
+
+        # call decoder
+        if self._metrics is None:
+            with self._cluster, tf1.device(self._cluster.GetPlacer()):
+                self._metrics = self._task.FPropDefaultTheta(decoder_inputs)
+
+        # compute loss gradient
+        loss = tf1.get_collection("per_loss")[0]
+        loss_gradient = tf1.gradients(loss, [x])[0]
+        return loss_gradient
+
+    def loss_gradient(self, x: np.ndarray, y: np.ndarray, batch_mode: bool = False, **kwargs) -> np.ndarray:
+        """
+        Compute the gradient of the loss function w.r.t. `x`.
+
+        :param x: Samples of shape `(nb_samples)`. Note that, it is allowable that sequences in the batch
+                  could have different lengths. A possible example of `x` could be:
+                  `x = np.ndarray([[0.1, 0.2, 0.1, 0.4], [0.3, 0.1]])`.
+        :param y: Target values of shape (nb_samples). Each sample in `y` is a string and it may possess different
+                  lengths. A possible example of `y` could be: `y = np.array(['SIXTY ONE', 'HELLO'])`.
+        :param batch_mode: If `True` calculate gradient per batch or otherwise per sequence.
+        :return: Loss gradients of the same shape as `x`.
+        """
+        if batch_mode:
+            gradients = self._loss_gradient_per_batch(x, y)
+        else:
+            gradients = self._loss_gradient_per_sequence(x, y)
+        return gradients
```

Comment:
We also need preprocessing for loss gradients, maybe something like this:
```suggestion
        """
        
        # Apply preprocessing
        x_preprocessed, y_preprocessed = self._apply_preprocessing(x, y, fit=False)
        
        if batch_mode:
            gradients = self._loss_gradient_per_batch(x_preprocessed, y_preprocessed)
        else:
            gradients = self._loss_gradient_per_sequence(x_preprocessed, y_preprocessed)
           
        gradients = self._apply_preprocessing_gradient(x, gradients) 
           
        return gradients
```